client/csi_endpoint.go

```diff
@@ -157,12 +158,50 @@ func (c *CSIController) DetachVolume(req *structs.ClientCSIControllerDetachVolum
 	return nil
 }
 
-func (c *CSIController) findControllerPlugin(name string) (csi.CSIPlugin, error) {
+// NodeDetachVolume is used to detach a volume from a CSI Cluster from
+// the storage node provided in the request.
+func (c *CSI) NodeDetachVolume(req *structs.ClientCSINodeDetachVolumeRequest, resp *structs.ClientCSINodeDetachVolumeResponse) error {
+	defer metrics.MeasureSince([]string{"client", "csi_node", "detach_volume"}, time.Now())
+
+	// The following block of validation checks should not be reached on a
+	// real Nomad cluster. They serve as a defensive check before forwarding
+	// requests to plugins, and to aid with development.
+	if req.PluginID == "" {
+		return errors.New("PluginID is required")
+	}
+	if req.VolumeID == "" {
```

Comment:
From right about here down, we need the volume namespace in case volumes with overlapping ids are deployed on the same node.

Reply:
From our sidebar discussion: these calls are specific to an allocation, which lives in a namespace already, so we're safe.